python/ray/ml/checkpoint.py

```diff
@@ -1,15 +1,51 @@
 import abc
-from typing import Dict, Optional
+import json
+import shutil
+import tarfile
+import tempfile
+
+import cloudpickle as pickle
+import os
+from typing import Any, Optional, Dict, Tuple
 
 import ray
+from ray.util.ml_utils.cloud import (
+    upload_to_bucket,
+    is_cloud_target,
+    download_from_bucket,
+)
+
+
+DICT_CHECKPOINT_FILE_NAME = "checkpoint.pkl"
+META_IS_DICT_CHECKPOINT = "is_dict_checkpoint"
 
 
 class Checkpoint(abc.ABC):
-    """Checkpoint interface"""
+    """Checkpoint interface.
+
+    This implementation provides interfaces to translate between
+    different checkpoint storage locations: Local FS storage, remote
+    node FS storage, data object, and cloud storage location.
 
-    def __init__(self, metadata: Optional[Dict] = None):
+    The following metadata keys are introduced for correct serialization
+    and deserialization of checkpoints:
+
+        is_dict_checkpoint (bool): If this is set, this checkpoint was
+            created from a dictionary and can thus be deserialized into a
+            dictionary checkpoint again for downstream processing.
+    """
+
+    def __init__(self, data: Any, metadata: Optional[Dict] = None):
+        self.data = data
```

Comment:
It would probably be more clear to store these as separate typed fields, like `self._dict: Optional[dict]` `self._uri: Optional[str]`, etc. and specify only one can be set at a time.